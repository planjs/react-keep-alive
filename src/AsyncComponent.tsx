import React from 'react';
import { bindLifecycleTypeName } from './utils/bindLifecycle';

interface IProps {
  setMounted: (value: boolean) => void;
  getMounted: () => boolean;
  onUpdate: () => void;
}

interface IState {
  component: any;
}

export default class AsyncComponent extends React.Component<IProps, IState> {
  state = {
    component: null,
  };

  /**
   * Force update child nodes
   *
   * @private
   * @returns
   * @memberof AsyncComponent
   */
  private forceUpdateChildren() {
    if (!this.props.children) {
      return;
    }
    const root: any = (this as any)._reactInternalFiber || (this as any)._reactInternalInstance;
    let node = root.child;
    let sibling = node;
    while (sibling) {
      while (true) {
        if (
          node.type &&
          node.type.displayName &&
          node.type.displayName.indexOf(bindLifecycleTypeName) !== -1
        ) {
          return;
        }
        if (node.stateNode) {
          break;
        }
        node = node.child;
      }
      if (typeof node.type === 'function') {
        node.stateNode.forceUpdate();
      }
      sibling = sibling.sibling;
    }
  }

  componentDidMount() {
    const { children } = this.props;
    Promise.resolve().then(() => this.setState({ component: children }));
  }

  componentDidUpdate() {
    this.props.onUpdate();
  }

  // Delayed update
  // In order to be able to get real DOM data
  shouldComponentUpdate() {
    if (!this.state.component) {
      // If it is already mounted asynchronously, you don't need to do it again when you update it.
      this.props.setMounted(false);
      return true;
    }
    Promise.resolve().then(() => {
      if (this.props.getMounted()) {
        this.props.setMounted(false);
        this.forceUpdateChildren();
        this.props.onUpdate();
      }
    });
    return false;
  }

  render() {
    return this.state.component;
  }
}
