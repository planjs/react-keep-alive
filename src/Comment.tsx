import React from 'react';
import ReactDOM from 'react-dom';
import noop from './utils/noop';

interface IReactCommentProps {
  onLoaded: () => void;
}

class ReactComment extends React.PureComponent<IReactCommentProps> {
  static defaultProps = {
    onLoaded: noop,
  };

  private parentNode!: Node;

  private currentNode!: Element;

  private commentNode!: Comment;

  private content!: string;

  componentDidMount() {
    const node = ReactDOM.findDOMNode(this) as Element;
    const commentNode = this.createComment();
    this.commentNode = commentNode;
    this.currentNode = node;
    this.parentNode = node.parentNode as Node;
    this.parentNode.replaceChild(commentNode, node);
    ReactDOM.unmountComponentAtNode(node);
    this.props.onLoaded();
  }

  componentWillUnmount() {
    this.parentNode.replaceChild(this.currentNode, this.commentNode);
  }

  private createComment() {
    let content = this.props.children;
    if (typeof content !== 'string') {
      content = '';
    }
    this.content = (content as string).trim();
    return document.createComment(this.content);
  }

  render() {
    return <div />;
  }
}

export default ReactComment;
