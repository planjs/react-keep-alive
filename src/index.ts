export { default as Provider } from './Provider';
export { default as KeepAlive } from './KeepAlive';
export { default as bindLifecycle } from './utils/bindLifecycle';
export { default as useKeepAliveEffect } from './utils/useKeepAliveEffect';
export { default as injectKeepAlive } from './injectKeepAlive';
