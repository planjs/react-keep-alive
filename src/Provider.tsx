import React from 'react';
import ReactDOM from 'react-dom';
import Comment from './Comment';
import KeepAliveContext from './contexts/KeepAliveContext';
import createEventEmitter from './utils/createEventEmitter';
import createUniqueIdentification from './utils/createUniqueIdentification';
import createStoreElement from './utils/createStoreElement';
import md5 from './utils/md5';

export const keepAliveProviderTypeName = '$$KeepAliveProvider';
export const START_MOUNTING_DOM = 'startMountingDOM';

export enum LIFECYCLE {
  MOUNTED,
  UPDATING,
  UNMOUNTED,
}

export interface ICacheItem {
  children: React.ReactNode;
  keepAlive: boolean;
  lifecycle: LIFECYCLE;
  renderElement?: HTMLElement;
  activated?: boolean;
  ifStillActivate?: boolean;
  reactivate?: () => void;
}

export interface ICache {
  [key: string]: ICacheItem;
}

export interface IKeepAliveProviderImpl {
  storeElement: HTMLElement;
  cache: ICache;
  keys: string[];
  eventEmitter: any;
  existed: boolean;
  providerIdentification: string;
  setCache: (identification: string, value: ICacheItem) => void;
  clearCache: (...keys: string[]) => void;
  clearAllCache: () => void;
  unactivate: (identification: string) => void;
  isExisted: () => boolean;
}

export interface IKeepAliveProviderProps {
  include?: string | string[] | RegExp;
  exclude?: string | string[] | RegExp;
  /**
   * 最大缓存组件数
   */
  max?: number;
  /**
   * 获取缓存组件存放位置
   */
  getStoreElementContainer?: () => HTMLElement;
}

export default class KeepAliveProvider
  extends React.PureComponent<IKeepAliveProviderProps>
  implements IKeepAliveProviderImpl {
  static displayName = keepAliveProviderTypeName;

  static defaultProps = {
    max: 10,
    getStoreElementContainer() {
      return createStoreElement();
    },
  };

  storeElement!: HTMLElement;

  // Sometimes data that changes with setState cannot be synchronized, so force refresh
  cache: ICache = Object.create(null);

  keys: string[] = [];

  eventEmitter = createEventEmitter();

  existed: boolean = true;

  private needRerender: boolean = false;

  providerIdentification: string = createUniqueIdentification();

  constructor(props: IKeepAliveProviderProps) {
    super(props);

    if (props.getStoreElementContainer) {
      this.storeElement = props.getStoreElementContainer();
    }
  }

  componentDidUpdate() {
    if (this.needRerender) {
      this.needRerender = false;
      this.forceUpdate();
    }
  }

  componentWillUnmount() {
    this.existed = false;
    document.body.removeChild(this.storeElement);
  }

  isExisted = () => this.existed;

  setCache = (identification: string, value: ICacheItem) => {
    const { cache, keys } = this;
    const { max } = this.props;
    const currentCache = cache[identification];
    if (!currentCache) {
      keys.push(identification);
    }
    this.cache[identification] = {
      ...currentCache,
      ...value,
    };
    this.forceUpdate(() => {
      // If the maximum value is set, the value in the cache is deleted after it goes out.
      if (currentCache) {
        return;
      }
      if (!max) {
        return;
      }
      const difference = keys.length - (max as number);
      if (difference <= 0) {
        return;
      }
      const spliceKeys = keys.splice(0, difference);
      this.forceUpdate(() => {
        spliceKeys.forEach((key) => {
          delete cache[key as string];
        });
      });
    });
  };

  unactivate = (identification: string) => {
    const { cache } = this;
    this.cache[identification] = {
      ...cache[identification],
      activated: false,
      lifecycle: LIFECYCLE.UNMOUNTED,
    };
    this.forceUpdate();
  };

  private startMountingDOM = (identification: string) => {
    this.eventEmitter.emit([identification, START_MOUNTING_DOM]);
  };

  clearCache = (...keys: string[]) => {
    keys.forEach((key) => this.unactivate(md5(`${this.providerIdentification}${key}`)));
  };

  clearAllCache = () => {
    Object.keys(this.cache).forEach(this.unactivate);
  };

  render() {
    const {
      cache,
      keys,
      providerIdentification,
      isExisted,
      setCache,
      clearCache,
      clearAllCache,
      existed,
      unactivate,
      storeElement,
      eventEmitter,
    } = this;
    const { children: innerChildren, include, exclude } = this.props;
    if (!storeElement) {
      return innerChildren;
    }
    return (
      <KeepAliveContext.Provider
        value={{
          cache,
          keys,
          existed,
          providerIdentification,
          isExisted,
          setCache,
          clearCache,
          clearAllCache,
          unactivate,
          storeElement,
          eventEmitter,
          include,
          exclude,
        }}
      >
        <>
          {innerChildren}
          {ReactDOM.createPortal(
            keys
              .map((identification) => {
                const currentCache = cache[identification];
                const { keepAlive, children, lifecycle } = currentCache;
                let cacheChildren = children;
                if (lifecycle === LIFECYCLE.MOUNTED && !keepAlive) {
                  // If the cache was last enabled, then the components of this keepAlive package are used,
                  // and the cache is not enabled, the UI needs to be reset.
                  cacheChildren = null;
                  this.needRerender = true;
                  currentCache.lifecycle = LIFECYCLE.UPDATING;
                }
                // current true, previous true | undefined, keepAlive false, not cache
                // current true, previous true | undefined, keepAlive true, cache

                // current true, previous false, keepAlive true, cache
                // current true, previous false, keepAlive false, not cache
                return cacheChildren ? (
                  <React.Fragment key={identification}>
                    <Comment>{identification}</Comment>
                    {cacheChildren}
                    <Comment onLoaded={() => this.startMountingDOM(identification)}>
                      {identification}
                    </Comment>
                  </React.Fragment>
                ) : null;
              })
              .filter((v) => v),
            storeElement,
          )}
        </>
      </KeepAliveContext.Provider>
    );
  }
}
