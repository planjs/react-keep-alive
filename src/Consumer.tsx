import React from 'react';
import Comment from './Comment';
import { LIFECYCLE, ICache, ICacheItem } from './Provider';

interface IConsumerProps {
  children: React.ReactNode;
  identification: string;
  keepAlive: boolean;
  cache: ICache;
  setCache: (identification: string, value: ICacheItem) => void;
  unactivate: (identification: string) => void;
}

class Consumer extends React.PureComponent<IConsumerProps> {
  private renderElement!: HTMLElement;

  private commentRef: any;

  private identification: string = this.props.identification;

  componentDidMount() {
    const { setCache, children, keepAlive } = this.props;
    this.renderElement = this.commentRef.parentNode;
    setCache(this.identification, {
      children,
      keepAlive,
      lifecycle: LIFECYCLE.MOUNTED,
      renderElement: this.renderElement,
      activated: true,
    });
  }

  componentDidUpdate() {
    const { setCache, children, keepAlive } = this.props;
    setCache(this.identification, {
      children,
      keepAlive,
      lifecycle: LIFECYCLE.UPDATING,
    });
  }

  componentWillUnmount() {
    const { unactivate } = this.props;
    unactivate(this.identification);
  }

  render() {
    const { identification } = this;
    return <Comment ref={(ref) => (this.commentRef = ref)}>{identification}</Comment>;
  }
}

export default Consumer;
