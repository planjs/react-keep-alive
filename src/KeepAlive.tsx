import React from 'react';
import AsyncComponent from './AsyncComponent';
import { START_MOUNTING_DOM, LIFECYCLE } from './Provider';
import keepAlive, { COMMAND } from './utils/keepAliveDecorator';
import changePositionByComment from './utils/changePositionByComment';

export interface KeepAliveProps<P = any> {
  key?: string;
  name?: string;
  disabled?: boolean;
  extra?: any;
  dynamicPros?: (componentProps: P, keepAliveProps: KeepAliveProps) => KeepAliveProps;
}

interface IKeepAliveInnerProps extends KeepAliveProps {
  _container: any;
}

class KeepAlive extends React.PureComponent<IKeepAliveInnerProps> {
  private bindUnmount: (() => void) | null = null;

  private bindUnactivate: (() => void) | null = null;

  private unmounted = false;

  private mounted = false;

  private ref: null | Element = null;

  private refNextSibling: null | Node = null;

  private childNodes: Node[] = [];

  componentDidMount() {
    const { _container } = this.props;
    const { notNeedActivate, identification, eventEmitter, keepAlive } = _container;
    notNeedActivate();
    const cb = () => {
      this.mount();
      this.listen();
      eventEmitter.off([identification, START_MOUNTING_DOM], cb);
    };
    eventEmitter.on([identification, START_MOUNTING_DOM], cb);
    if (keepAlive) {
      this.componentDidActivate();
    }
  }

  componentDidActivate() {
    // tslint-disable
  }

  componentDidUpdate() {
    const { _container } = this.props;
    const { notNeedActivate, isNeedActivate } = _container;
    if (isNeedActivate()) {
      notNeedActivate();
      this.mount();
      this.listen();
      this.unmounted = false;
      this.componentDidActivate();
    }
  }

  componentWillUnactivate() {
    this.unmount();
    this.unListen();
  }

  componentWillUnmount() {
    if (!this.unmounted) {
      this.unmounted = true;
      this.unmount();
      this.unListen();
    }
  }

  private mount() {
    const {
      _container: { cache, identification, storeElement, setLifecycle },
    } = this.props;
    this.setMounted(true);
    const { renderElement } = cache[identification];
    setLifecycle(LIFECYCLE.UPDATING);
    changePositionByComment(identification, renderElement, storeElement);
  }

  private correctionPosition = () => {
    if (this.ref && this.ref.parentNode && this.ref.nextSibling) {
      const childNodes = this.ref.childNodes as any;
      this.refNextSibling = this.ref.nextSibling;
      this.childNodes = [];
      while (childNodes.length) {
        const child = childNodes[0];
        this.childNodes.push(child);
        this.ref.parentNode.insertBefore(child, this.ref.nextSibling);
      }
      this.ref.parentNode.removeChild(this.ref);
    }
  };

  private retreatPosition = () => {
    if (this.ref && this.refNextSibling && this.refNextSibling.parentNode) {
      for (const child of this.childNodes) {
        this.ref.appendChild(child);
      }
      this.refNextSibling.parentNode.insertBefore(this.ref, this.refNextSibling);
    }
  };

  private unmount() {
    const {
      _container: { identification, storeElement, cache, setLifecycle },
    } = this.props;
    const { renderElement, ifStillActivate, reactivate } = cache[identification];
    setLifecycle(LIFECYCLE.UNMOUNTED);
    this.retreatPosition();
    changePositionByComment(identification, storeElement, renderElement);
    if (ifStillActivate) {
      reactivate();
    }
  }

  private listen() {
    const {
      _container: { identification, eventEmitter },
    } = this.props;
    eventEmitter.on(
      [identification, COMMAND.CURRENT_UNMOUNT],
      (this.bindUnmount = this.componentWillUnmount.bind(this)),
    );
    eventEmitter.on(
      [identification, COMMAND.CURRENT_UNACTIVATE],
      (this.bindUnactivate = this.componentWillUnactivate.bind(this)),
    );
  }

  private unListen() {
    const {
      _container: { identification, eventEmitter },
    } = this.props;
    eventEmitter.off([identification, COMMAND.CURRENT_UNMOUNT], this.bindUnmount);
    eventEmitter.off([identification, COMMAND.CURRENT_UNACTIVATE], this.bindUnactivate);
  }

  private setMounted = (value: boolean) => {
    this.mounted = value;
  };

  private getMounted = () => this.mounted;

  render() {
    // The purpose of this div is to not report an error when moving the DOM,
    // so you need to remove this div later.
    return (
      <div ref={(ref) => (this.ref = ref)}>
        <AsyncComponent
          setMounted={this.setMounted}
          getMounted={this.getMounted}
          onUpdate={this.correctionPosition}
        >
          {this.props.children}
        </AsyncComponent>
      </div>
    );
  }
}

export default keepAlive<KeepAliveProps>(KeepAlive);
