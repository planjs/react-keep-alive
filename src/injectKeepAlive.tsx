import React from 'react';
import KeepAlive, { KeepAliveProps } from './KeepAlive';

export default function injectKeepAlive(opt: KeepAliveProps) {
  return function <P = {}>(Component: React.ComponentType<P>) {
    return (props: P) => (
      <KeepAlive {...opt}>
        <Component {...props} />
      </KeepAlive>
    );
  };
}
