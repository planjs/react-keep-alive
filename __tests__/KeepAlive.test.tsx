import React from 'react';
import {
  render,
  getByTestId,
  cleanup,
  getNodeText,
  fireEvent,
  waitFor,
  getByText,
} from '@testing-library/react';

import { Provider, KeepAlive } from '../src';

const Counter: React.FC = () => {
  const [count, setCount] = React.useState(0);

  const handleAdd = React.useCallback(() => {
    setCount(count + 1);
  }, [count]);

  return (
    <div>
      <div data-testid="count">{count}</div>
      <button data-testid="counterAdd" onClick={handleAdd}>
        +
      </button>
    </div>
  );
};

const App: React.FC = () => {
  const [show, setShow] = React.useState(true);

  const handleToggleShow = React.useCallback(() => {
    setShow(!show);
  }, [show]);

  return (
    <Provider>
      <div>
        {show}
        {show ? (
          <KeepAlive key="counter">
            <Counter />
          </KeepAlive>
        ) : null}
        <button data-testid="toggle" onClick={handleToggleShow}>
          toggle
        </button>
      </div>
    </Provider>
  );
};

test('Test <KeepAlive />', async () => {
  afterEach(cleanup);
  const { container } = render(<App />);
  await waitFor(() => getByTestId(container, 'count'));

  const getCount = () => getByTestId(container, 'count');
  // init state
  expect(container).toMatchSnapshot();
  expect(getNodeText(getCount())).toBe('0');

  // add count
  const counterAdd = getByTestId(container, 'counterAdd');
  fireEvent.click(counterAdd);
  expect(container).toMatchSnapshot();
  expect(getNodeText(getCount())).toBe('1');

  // hide Counter
  const toggleBtn = getByTestId(container, 'toggle');
  fireEvent.click(toggleBtn);
  try {
    await waitFor(() => getByTestId(container, 'count'));
  } catch (e) {
    expect(e.message).toMatch('Unable to find an element');
    expect(container).toMatchSnapshot();
  }

  // show Counter
  fireEvent.click(toggleBtn);
  expect(container).toMatchSnapshot();
  expect(getNodeText(getCount())).toBe('1');
});
