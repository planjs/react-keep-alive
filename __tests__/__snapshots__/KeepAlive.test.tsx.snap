// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <KeepAlive /> 1`] = `
<div>
  <div>
    <!--keep-alive-af4509-->
    <div>
      <div
        data-testid="count"
      >
        0
      </div>
      <button
        data-testid="counterAdd"
      >
        +
      </button>
    </div>
    <!--keep-alive-af4509-->
    <button
      data-testid="toggle"
    >
      toggle
    </button>
  </div>
</div>
`;

exports[`Test <KeepAlive /> 2`] = `
<div>
  <div>
    <!--keep-alive-af4509-->
    <div>
      <div
        data-testid="count"
      >
        1
      </div>
      <button
        data-testid="counterAdd"
      >
        +
      </button>
    </div>
    <!--keep-alive-af4509-->
    <button
      data-testid="toggle"
    >
      toggle
    </button>
  </div>
</div>
`;

exports[`Test <KeepAlive /> 3`] = `
<div>
  <div>
    <button
      data-testid="toggle"
    >
      toggle
    </button>
  </div>
</div>
`;

exports[`Test <KeepAlive /> 4`] = `
<div>
  <div>
    <!--keep-alive-af4509-->
    <div>
      <div>
        <div
          data-testid="count"
        >
          1
        </div>
        <button
          data-testid="counterAdd"
        >
          +
        </button>
      </div>
    </div>
    <!--keep-alive-af4509-->
    <button
      data-testid="toggle"
    >
      toggle
    </button>
  </div>
</div>
`;
